### ISSUE TITLE
Warn when writing *args after a kwarg in calls

### ISSUE THREAD

**Issue by Gouvernathor** [microsoft/pyright#5755]
The `(kw=val, *args)` syntax **_in calls_** (**not** in signatures) is misleading : if you use it because you think it changes anything compared to `(*args, kw=val)` then you're wrong about how python calls actually work.
That syntax was only kept in python (in and after [PEP 448](https://peps.python.org/pep-0448/)) due to backwards-compatibility concerns, but the interpreter internally does the exact same tweak I described (an element of [proof](https://docs.python.org/3/library/ast.html#ast.Call)), and its peculiarity has lead Python core devs to push to deprecate it.
Even if that doesn't end up happening, it should be warned against.

The issues it raises are, first, that the syntax breaks the usual order in which arguments are evaluated, since the `args` expression will be evaluated before the `val` expression. It's otherwise done strictly in a left-to-right fashion.
Second, it can lead people to believe that the `kw` parameter, if it's not a keyword-only parameter in the original signature, will be matched with `val` if it's the first parameter of the called object. Buit in fact, that will only work if `args` is empty, and fail otherwise (because `kw` will be double-bound). That will most often happen in super calls or generally in proxy calls.

Additional sources :
https://mail.python.org/archives/list/python-dev@python.org/thread/I6AUYV6DVEMP7XVYVDWC62N6PK6FHX3H/ https://discuss.python.org/t/rationale-for-supporting-foo-kw-1-args-keyword-arguments-before-positional-unpackings/26073/7

**Comment by erictraut** [comment]
I don't think it makes sense for a type checker to emit a diagnostic for this pattern in general. It's a legal form with well-defined and deterministic behaviors. If you want to disallow this pattern in your code, that's more of a stylistic decision and a custom linter rule would be an appropriate way to enforce this.

Pyright already emits errors when there is an error matching arguments to parameters. (It also obviously emits errors where such matches are possible but the corresponding types are incompatible.)

I'm not sure I fully understand some of what you're saying above. For example, you said `the syntax breaks the usual order in which arguments are evaluated`. In my experiments, this doesn't appear to be the case.

```python
def f(a=None, b=None):
    print(f"f received a={a}, b={b}")
    return a, b

def eval(val: int):
    print(f"eval {val}")
    return val

f(eval(1), b=eval(2))
f(b=eval(2), *[eval(1)])
```

This code sample demonstrates that the argument expressions are evaluated in left-to-right ordering in both cases, as one would expect. The mapping of arguments to parameters isn't done in a strictly left-to-right manner, but that's generally not true when whenever keyword arguments are involved.

Perhaps you can provide a concrete example where pyright doesn't currently emit an error but there is a runtime exception?

**Comment by Gouvernathor** [comment]
In your example, "eval 1" is printed before "eval 2" in both cases even though the second instance shows "1" being written on the right of "2", so the arguments of the call are indeed not evaluated in left-to-right order in the second case. And my initial point is that there is no other situation where arguments are not evaluated in strict left-to-right order.
![image](https://github.com/microsoft/pyright/assets/44340603/040e30c0-d4c9-4465-af73-dc08e7112302)
I don't know why you conclude that "the argument expressions are evaluated in left-to-right ordering in both cases" since it's visibly not what happens in the second case. Unless you have a different result on your end of course ? but I would be surprised.
Maybe you're mistaking the order of arguments with the order of parameters of the callable ? but I'd also be surprised since the order of parameters has nothing to do with the evaluation order of the arguments.

A common case where such a mistake can be made is the following example, a wrapper of [shutil.copytree](https://docs.python.org/3/library/shutil.html#shutil.copytree) :
```py
def copytree_to_there(src, *args, **kwargs):
    return shutil.copytree(src, dst=some_predefined_path, *args, **kwargs)
```
In this example, passing any positional argument other than `src` will fail, because the `dst` parameter will be double-bound. Yet from the looks of the call, it may seem like dst goes in second place, before the other positionals. But in fact the call only works when args is empty. Forcing `(src, *args, dst=prederfined_path, **kwargs)`

Maybe pyright/pylance is not the best place to include this check, though. I thought pyright emitted a warning for `a = 5,` and advised adding parentheses around the tuple, but it doesn't seem to, I must be mistaking with another linter. That was an example of legal behavior worth warning against I had in mind.

**Comment by erictraut** [comment]
Ah, I see what you mean. Yes, that is surprising.

Here's a simple and concrete example of where this fails in a surprising way at runtime.

```python
f((x := 1), b=x + 1)
f(b=(y := 1), *[y + 1])  # Fails at runtime
f(b=z + 1, *[(z := 1)])  # Works at runtime, but pyright displays error
```

The fact that pyright's behavior doesn't match the runtime is something that I would like to fix if it's feasible. I'll investigate that further.

As for emitting a diagnostic for this pattern in general, I'll think about that a bit more, but my initial reaction is that this is more of a linter feature, and pyright generally tries to stick to type checking.

**Comment by erictraut** [comment]
I've pyright's logic to mirror the evaluation order for arguments at runtime. 

I'm going to refrain from adding a general diagnostic for now. We could reconsider this in the future based on feedback from pyright and pylance users. To date, no one else has asked for this functionality, and I don't think it's a very common source of bugs in general.

**Comment by erictraut** [comment]
This is included in pyright 1.1.324, which I just published. It will also be included in a future release of pylance.
